File: nQueensProblem.java
class Solution {
    List<List<String>> result;
    int[][] board;
    int m;
    public List<List<String>> solveNQueens(int n) {
        result = new ArrayList<>();
        board = new int[n][n];
        m = n;
        
        if(n == 0 ) return result;
        
        backtrack(0);
        return result;
    }
    
    private void backtrack(int row){
        List<String> li = new ArrayList<>();
        
        //base
        if(row == m ){ //placed all queens
        for(int i = 0 ; i < m ; i++){
            StringBuilder sb = new StringBuilder();
            for(int j = 0 ; j <m ; j++){
                if(board[i][j] == 0){
                    sb.append('.');
                }
                else{
                    sb.append('Q');
                }
            }
        li.add(sb.toString());    
        }
            result.add(li);
            return;
        }
        
        //logic
        for(int i = 0 ; i < m ; i ++){
           if(isSafe(row,i)){
            //action
            board[row][i] = 1;
            //recurse
            backtrack(row+1);
            //backtrack
            board[row][i] = 0;
        }
    }
    }
    
    private boolean isSafe(int r, int c){
        //column above
       for(int k=0;k < r; k++){
            if(board[k][c] == 1) return false;
        }
    
        //diagonal above up left
         int i = r-1; int j = c-1;
        while(i>=0 && j>=0){
            if(board[i][j] == 1) return false;
                i--;j--;
        }
        //diagonal above up right
        int iNew = r-1; int jNew = c+1;
        while(iNew >= 0 && jNew < m){
            if(board[iNew][jNew] == 1) return false;
                iNew-- ; jNew++;
        }
    return true;
}
}

//Time : Nchoices in 1st row,N-2 choices in 2nd row ..... hence O(n!)
//Space: O(nxn) board space + O(n) recursive space + O(n) string builder aymtotically O(nxn)

File: n-queensProblem.py
'''
Remember: If I am able to place the queen into a particular row,
1. Abandon the exceution for that row i.e. parent row inside the recursion-stack and over
to the next row
2. If not able to place the queen anywhere into the row, recursively roll-back to the parent 
row, rollback to the previous value for the particular cell and continue
'''

'''
Time Complexity  : factorial(n)
Space Complexity : N*N matrix initialization 0(n^2)
                   recursive stack 0(n)
On Leet-code: Yes
'''
class Solution:
    
    def __init__(self):
        self.resultList = []
    
    def canBePlaced(self,board,row,col):
        
        # chk "Q" at the top
        itr_r = row
        while itr_r >= 0:
            if board[itr_r][col] == "Q":
                return False
            itr_r -= 1
        
        # chk "Q" at the top-right
        itr_r = row
        itr_c = col
        while itr_r >= 0 and itr_c < len(board[row]):
            if board[itr_r][itr_c] == "Q":
                return False
            itr_r -= 1
            itr_c += 1
        
        # chk "Q" at the top-left
        itr_r = row
        itr_c = col
        while itr_r >= 0 and itr_c >= 0:
            if board[itr_r][itr_c] == "Q":
                return False
            itr_r -= 1
            itr_c -= 1
        
        # default return
        return True        
    
    def placeQueen(self,board,row):
        
        # base-case
        if row == len(board):
            
            # create a tempList
            tempList = [None] * len(board)
            
            # I have my n-queens in non-attacking formation
            for i in range(0,len(board)):
                tempStr = ''
                for j in range(0,len(board[i])):
                    if board[i][j] == "Q":
                        tempStr += 'Q'
                    else:
                        tempStr += '.'
                
                # my tempStr is ready!!
                tempList[i] = str(tempStr)
            
            # append the tempList into the resultList
            self.resultList.append(tempList)
            return
                
        # logic-case
        # Iterate the row
        for col in range(0,len(board)):
            flag = self.canBePlaced(board,row,col)
            
            # chk flag result and decide
            if flag == True:
                # I can place -- action
                board[row][col] = "Q"
                
                # do recursive call -- recursive
                self.placeQueen(board,row+1)
                
                # rollback to previous state
                board[row][col] = None
            
            # if flag == False; go-to the next col
        
        # exited for-loop; go-back to the parent in recursive stack
        return
        
    def solveNQueens(self, n: int) -> List[List[str]]:
        
        # initialize a board
        board = [[None for col in range(0,n)] for row in range(0,n)]
        
        # call the function
        self.placeQueen(board,0)
        
        # check the result
        # print(self.resultList)
        return self.resultList